internal/controllers/machine/machine_controller_status.go

```diff
@@ -512,6 +515,102 @@ func (c machineConditionCustomMergeStrategy) Merge(conditions []v1beta2condition
 	).Merge(conditions, conditionTypes)
 }
 
+// transformControlPlaneAndEtcdConditions Group readiness gates for control plane conditions and etcd when they have the same messages.
+// Note: the implementation is based on KCP conditions, but ideally other control plane implementation could
+// take benefit from this optimization by naming conditions with APIServer, ControllerManager, Scheduler or the Etcd prefix.
+func transformControlPlaneAndEtcdConditions(messages []string) []string {
+	isControlPlaneCondition := func(c string) bool {
+		if strings.HasPrefix(c, "* APIServer") {
+			return true
+		}
+		if strings.HasPrefix(c, "* ControllerManager") {
+			return true
+		}
+		if strings.HasPrefix(c, "* Scheduler") {
+			return true
+		}
+		// Note. Etcd pod healthy is considered as part of control plane components in KCP
+		// because it is not part of the etcd cluster.
+		// Might be in future we want to make this check more strictly tight to KCP machines e.g. by checking the machine's owner;
+		// for now, we consider checking for the exact condition name as an acceptable trade off (same below).
+		if c == "* EtcdPodHealthy" {
+			return true
+		}
+		return false
+	}
+
+	out1 := []string{}
+	controlPlaneConditionsCount := 0
+	controlPlaneMsg := ""
+	for _, m := range messages {
+		sep := strings.Index(m, ":")
+		if sep == -1 {
+			out1 = append(out1, m)
+			continue
+		}
+
+		c, msg := m[:sep], m[sep+1:]
+		if !isControlPlaneCondition(c) {
+			out1 = append(out1, m)
+			continue
+		}
+
+		if controlPlaneMsg == "" {
+			controlPlaneMsg = msg
+			out1 = append(out1, fmt.Sprintf("* Control plane components:%s", msg))
+			controlPlaneConditionsCount++
+			continue
+		}
+
+		if controlPlaneMsg != msg {
+			out1 = messages
+			break
+		}
+		controlPlaneConditionsCount++
+	}
+	if controlPlaneConditionsCount == 1 {
+		out1 = messages
+	}
+
+	isEtcdCondition := func(c string) bool {
+		return c != "* EtcdPodHealthy" && strings.HasPrefix(c, "* Etcd")
+	}
+
+	out2 := []string{}
+	etcdConditionsCount := 0
+	etcdMemberMsg := ""
```

Comment:
This is not a real use case right? At least none that happens with KCP.

If I understand all of this correctly we are grouping all "* Etcd" conditions except EtcdPodHealthy into "* Etcd member:%"?

1. Is there a use case that we are aware of where we need this?
2. If not, how do we know it's safe to assume that Etcd* conditions all belong to Etcd member

To be honest, if there is no known use case for it. I would drop everything below here

Reply:
Removed, we can always add it back if we need it